### PR TITLE
Initial code for decorators and optional naming of workflows and activities

### PR DESCRIPTION
**Pull request by mukundansundar**
# Description

- When a decorator is added for an activity or workflow, it is registered against the instance through which the decorator is accessed.
- For optional naming, I followed a slightly modified suggestion made in this [issue](https://github.com/dapr/python-sdk/issues/623). I add an `_alternate_name` for each `workflow`, `activity` function/method and when a new name a given, the `alternate_name` which is the user given name takes precedence over the `__name__` of the function/method. 
- Pending TODOs:
  - [x] Adding unit tests as needed for workflow/activity decorator and optional name feature. 
  - [ ] The serialize/deserializer addition will be done as a separate PR. A separate issue will be created to track it. 

Following through with the discussion in #575, I have added partial validation to the `chaining` and `fan out fan in` examples alone as the other two require user input and take a longer time to complete also. 

@berndverst  and @cgillum  For now I have left the `register_workflow` and the `register_activity` methods as is without any changes, so users can use either a decorator patter or use `register_workflow` and `register_activity` for now. 

So if we continue to support both, there has to be the case that any parameter or functionality added as part of the decorator will in essence be added as part of the `register_*` methods, that is the code flow today also. 

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

part of #635 
closes https://github.com/dapr/python-sdk/issues/623

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: ext/dapr-ext-workflow/tests/test_workflow_runtime.py [hunk 1]
```diff
@@ -34,20 +34,56 @@ def add_named_activity(self, name: str, fn):
...
+        assert listOrchestrators == wanted_orchestrator
+        assert client_wf._registered_name == self.mock_client_wf.__name__
+
+        client_activity = (self.runtime_options.activity())(self.mock_client_activity)
+        wanted_activity = [self.mock_client_activity.__name__]
+        assert listActivities == wanted_activity
+        assert client_activity._registered_name == self.mock_client_activity.__name__
+
+    def test_both_decorator_and_register(self):
+        client_wf = (self.runtime_options.workflow(name="test_wf"))(self.mock_client_wf)
```
Comment (mukundansundar) [inline ext/dapr-ext-workflow/tests/test_workflow_runtime.py:70]: When called with the `@` operator, the function object is replaced in place with the wrapped function only.

**Comment by bkcsfi** [comment]
Hi,

I like where this is going but I have some suggestions that would make it easier for my use case.

I think it would be helpful to offer a decorator function that sets the function `_registered_name` attribute independent of having to instantiate a WorkflowRuntime. 

All of our code explicitly passes an already instantiated  WorkflowRuntime instance from the top-most level to registration functions in lower-level modules. This way we only instantiate a WorkflowRuntime once, at the top level, and not on every module import. Being able to set `_registered_name` separately from the registration process would match our current architecture.


Then  `register_workflow()` and `register_activity()` could be adjusted to detect when the passed function already has a `_registered_name` attribute and use that if the `name` parameter had not been set. If neither `name` is passed nor `_registered_name` is found, then default to function `__name__`

**Comment by mukundansundar** [comment]
> Hi,
> 
> I like where this is going but I have some suggestions that would make it easier for my use case.
> 
> I think it would be helpful to offer a decorator function that sets the function `_registered_name` attribute independent of having to instantiate a WorkflowRuntime.
> 
> All of our code explicitly passes an already instantiated WorkflowRuntime instance from the top-most level to registration functions in lower-level modules. This way we only instantiate a WorkflowRuntime once, at the top level, and not on every module import. Being able to set `_registered_name` separately from the registration process would match our current architecture.
> 
> Then `register_workflow()` and `register_activity()` could be adjusted to detect when the passed function already has a `_registered_name` attribute and use that if the `name` parameter had not been set. If neither `name` is passed nor `_registered_name` is found, then default to function `__name__`

I believe that the name a function registers with for a workflow should be provided only during the register call or be the name of the function itself. I do not think we should add a separate decorator for adding a `_registrered_name` attribute.

Instead is there any specific scenario where the optional name cannot be set with the `register_workflow()` or `register_activity()` method itself?

**Comment by mukundansundar** [comment]
@bkcsfi I have modified to add an `alternate_name` decorator which adds an `_alternate_name` attribute to the function which can be used instead of `__name__` ... Please let us know if this works for your use case.

File: examples/workflow/fan_out_fan_in.py [hunk 2]
```diff
@@ -59,4 +57,4 @@ def process_results(ctx, final_result: int):
     print(f'Workflow started. Instance ID: {instance_id}')
     state = wf_client.wait_for_workflow_completion(instance_id)
 
-    workflowRuntime.shutdown()
+    wfr.shutdown()
```
Comment (cgillum) [inline examples/workflow/fan_out_fan_in.py:60]: This can probably be done in another PR if you'd prefer, but one of the improvements we discussed was making it so that the runtime can be stopped automatically if we allow the `start()` method to be invoked using a [`with`](https://peps.python.org/pep-0343/) block.

File: examples/workflow/README.md [hunk 3]
```diff
@@ -22,10 +22,20 @@ Each of the examples in this directory can be run directly from the command line
 ### Task Chaining
 
 This example demonstrates how to chain "activity" tasks together in a workflow. You can run this sample using the following command:
+<!--STEP
```
Comment (cgillum) [inline examples/workflow/README.md:25]: I appreciate you adding the mechanical markdown to this example!

File: examples/workflow/fan_out_fan_in.py [hunk 4]
```diff
@@ -14,7 +14,9 @@
 from typing import List
 import dapr.ext.workflow as wf
 
+wfr = wf.WorkflowRuntime("localhost", "50001")
```
Comment (cgillum) [inline examples/workflow/fan_out_fan_in.py:17]: Should we instead rely on the default values for this? Otherwise there could be problems if someone runs dapr and sets a non-standard gRPC port.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py [hunk 5]
```diff
@@ -80,3 +107,141 @@ def start(self):
...
+        This example shows how to register a workflow function with a name:
+
+                from dapr.ext.workflow import WorkflowRuntime
+                wfr = WorkflowRuntime()
+
+                @wfr.workflow(name="add")
+                def add(ctx, x: int, y: int) -> int:
+                    return x + y
+
+        This example shows how to register a workflow function without a name:
```
Comment (cgillum) [inline ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:123]: It's not that it registers it without a name, but that it uses the function name as the workflow name.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py [hunk 6]
```diff
@@ -80,3 +107,141 @@ def start(self):
     def shutdown(self):
         """Stops the listening for work items on a background thread."""
         self.__worker.stop()
+
+    def workflow(self, __fn: Workflow = None, *, name: Optional[str] = None):
+        """Decorator to register a workflow function.
+
+        This example shows how to register a workflow function with a name:
```
Comment (cgillum) [inline ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:114]:
```suggestion
        This example shows how to register a workflow function with an explicit name:
```

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py [hunk 7]
```diff
@@ -80,3 +107,141 @@ def start(self):
...
+                    from dapr.ext.workflow import WorkflowRuntime
+                    wfr = WorkflowRuntime()
+
+                    @wfr.workflow
+                    def add(ctx, x: int, y: int) -> int:
+                        return x + y
+
+        Args:
+            name (Optional[str], optional): Name to identify the workflow function as in
+            the workflow runtime. Defaults to None.
```
Comment (cgillum) [inline ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:134]:
```suggestion
            the workflow runtime. Defaults to the name of the function.
```

File: ext/dapr-ext-workflow/dapr/ext/workflow/dapr_workflow_context.py [hunk 8]
```diff
@@ -53,6 +53,10 @@ def create_timer(self, fire_at: Union[datetime, timedelta]) -> task.Task:
 
     def call_activity(self, activity: Callable[[WorkflowActivityContext, TInput], TOutput], *,
                       input: TInput = None) -> task.Task[TOutput]:
+        if hasattr(activity, '_alternate_name'):
+            return self.__obj.call_activity(activity=activity.__dict__['_alternate_name'],
+                                            input=input)
+        # this return should ideally never execute
```
Comment (cgillum) [inline ext/dapr-ext-workflow/dapr/ext/workflow/dapr_workflow_context.py:59]: Why is this "ideal"? Isn't this the case where the user chose not to use the decorator, which we plan to allow?

**Review by cgillum — COMMENTED** [review]
A few small comments.

File: examples/workflow/fan_out_fan_in.py (hunk 2 above)
Comment (mukundansundar) [reply examples/workflow/fan_out_fan_in.py:60]: Will leave it for another PR.

File: examples/workflow/fan_out_fan_in.py (hunk 4 above)
Comment (mukundansundar) [reply examples/workflow/fan_out_fan_in.py:17]: Will change it.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py (hunk 6 above)
Comment (mukundansundar) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:114]: changed it to `alternate` name

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py (hunk 7 above)
Comment (mukundansundar) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:134]: In this case, it does default to `None`, but is then overridden by the function name.

File: ext/dapr-ext-workflow/dapr/ext/workflow/dapr_workflow_context.py (hunk 8 above)
Comment (mukundansundar) [reply ext/dapr-ext-workflow/dapr/ext/workflow/dapr_workflow_context.py:59]: No in both scenarios, the attribute `_alternate_name` will be set for the function being called. So ideally the if clause will always execute.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py (hunk 7 above)
Comment (cgillum) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:135]: I guess it depends on whether this description is intended to describe the function signature of the what the user can expect - i.e., the semantic behavior.

**Review by cgillum — APPROVED** [review]
I'm good with these changes. 👍🏽

**Comment by bkcsfi** [comment]
@mukundansundar Hi thank you for revising the decorator and register functions. The changes meet all my needs.

Do you have an existing issue or discussion for `The serialize/deserializer addition will be done as a separate PR. A separate issue will be created to track it.` ?   

I am using some ugly approaches to try to keep activity and workflow  'input deseralization' and 'output serialization' tied to activity and workflow function declarations, rather than forcing the caller to do that work. 

I would appreciate seeing what you're planning there, as well as maybe getting durable task to `send()` deserialized activity results to the workflow..

Thanks,
-Brad

**Comment by mukundansundar** [comment]
> @mukundansundar Hi thank you for revising the decorator and register functions. The changes meet all my needs.
> 
> Do you have an existing issue or discussion for `The serialize/deserializer addition will be done as a separate PR. A separate issue will be created to track it.` ?
> 
> I am using some ugly approaches to try to keep activity and workflow 'input deseralization' and 'output serialization' tied to activity and workflow function declarations, rather than forcing the caller to do that work.
> 
> I would appreciate seeing what you're planning there, as well as maybe getting durable task to `send()` deserialized activity results to the workflow..
> 
> Thanks, -Brad

I will be creating an issue following a comment/discussion from you in discord.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py [hunk 9]
```diff
@@ -50,17 +52,30 @@ def __init__(self, host: Optional[str] = None, port: Optional[str] = None):
...
             daprWfContext = DaprWorkflowContext(ctx)
             if inp is None:
                 return fn(daprWfContext)
             return fn(daprWfContext, inp)
 
-        self.__worker._registry.add_named_orchestrator(fn.__name__, orchestrationWrapper)
-
-    def register_activity(self, fn: Activity):
+        if hasattr(fn, '_workflow_registered'):
+            # whenever a workflow is registered, it also has _alternate_name attribute
```
Comment (DeepanshuA) [inline ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:64]:
```suggestion
            # whenever a workflow is already registered, it also has an existing _alternate_name attribute
```

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py [hunk 10]
```diff
@@ -71,7 +86,19 @@ def activityWrapper(ctx: task.ActivityContext, inp: Optional[TInput] = None):
                 return fn(wfActivityContext)
             return fn(wfActivityContext, inp)
 
-        self.__worker._registry.add_named_activity(fn.__name__, activityWrapper)
+        if hasattr(fn, '_activity_registered'):
+            # whenever an activity is registered, it also has _alternate_name attribute
```
Comment (DeepanshuA) [inline ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:90]:
```suggestion
            # whenever an activity is already registered, it also has an existing _alternate_name attribute
```

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py [hunk 11]
```diff
@@ -50,17 +52,30 @@ def __init__(self, host: Optional[str] = None, port: Optional[str] = None):
         self.__worker = worker.TaskHubGrpcWorker(host_address=uri.endpoint, metadata=metadata,
                                                  secure_channel=uri.tls)
 
-    def register_workflow(self, fn: Workflow):
+    def register_workflow(self, fn: Workflow, *, name: Optional[str] = None):
```
Comment (DeepanshuA) [inline ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:55]: Is it possible to add a small doc section somewhere in the python-sdk repo itself, where-in we briefly describe that re-registering (or mixing the registration methods to register multiple times) will not succeed?

Comment (bkcsfi) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:55]: I had previously suggested that registering a workflow/activity with an explicit name argument that already has an assigned _alternate_name should raise an exception.

But now I wonder if there is a legitimate use-case for overriding a previously set `_alternate name`  during the registration step.  

Is it likely that activities and workflows might be packaged into a "python library", and the consumer of that library might want to override the alternate name during the registration process? 

Perhaps to have precise control over versioning so that existing workflows won't break?

Should the register function log a warning instead of raising an exception,  and then override the previously assigned `_alternate_name` with the passed in `name`?

I do think the decorator should raise an exception if `_alternate_name` is already set on the function. 

Finally, should the `_alternate_name` attribute be renamed to `_dapr_alternate_name`  just to namespace it a bit more?

Comment (mukundansundar) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:55]: I think making the change from `_alternate_name` to`_dapr_alternate_name` is fine, I will make that change. 
On the other parts, I do not think for this PR we need to allow multiple registrations of a workflow with different names. Let's open a separate feature request and discuss the pros, cons of that feature.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py (hunk 10 above)
Comment (mukundansundar) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py:90]: changed.

File: ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py (hunk 9 above)
Comment (mukundansundar) [reply ext/dapr-ext-workflow/dapr/ext/workflow/workflow_runtime.py]: changed

**Review by berndverst — APPROVED** [review]
